components/esp32/cpu_start.c

```diff
@@ -332,7 +332,7 @@ void start_cpu0_default(void)
     intr_matrix_clear();
 
 #ifndef CONFIG_CONSOLE_UART_NONE
-#ifdef CONFIG_PM_ENABLE
+#if defined(CONFIG_PM_ENABLE) && CONFIG_CONSOLE_UART_BAUDRATE <= REF_CLK_FREQ
```

Comment:
Hi @KonstantinKondrashov -

A couple of notes:
 - One thing I wasn't sure of is if REF_CLK_FREQ can change - e.g. if it's possible to configure this in the hardware to be something other than 1MHz (I couldn't find much about this in the reference manual).  The one thing I don't like about placing it inside Kconfig in this way is it doesn't reference the constant for the reference clock frequency.  The code could instead have a #error that references the constant for REF_CLK_FREQ instead of placing the constant reference inside the configuration menu.
 - Arguably it's only a problem based on CONFIG_PM_DFS_INIT_AUTO being set, so if it were done inside Kconfig it should be dependent on that, not on PM_ENABLE
 - Another thing I wasn't sure of was if there was a maximum ratio of baud to input clock rate.  E.g. I don't think the UART runs well off of REFCLK at 1MHz.  I couldn't find information about this in the documentation for the chip either.  For other chips, I know the baud rate cannot be more than 1/8th of the input clock rate into the module for instance.

-- Tim

Reply:
When PM_ENABLE is enabled, UART will be clocked from REF_TICK clock, which is 1MHz. We should not give the opportunity a customer to set CONSOLE_UART_BAUDRATE more than possible. We need to add the limitation in `cpu_ctart.c` file as well:
```
#ifdef CONFIG_PM_ENABLE
    _Static_assert(CONFIG_CONSOLE_UART_BAUDRATE < REF_CLK_FREQ, "The CONFIG_CONSOLE_UART_BAUDRATE option have to be less REF_CLK_FREQ (1MHz)");
...
```
As for _the maximum ratio of baud to input clock rate_ we can be referred to the implementation of [uart_set_baudrate()](https://github.com/espressif/esp-idf/blob/master/components/driver/uart.c#L201-L213) function. If UART has been configured to use REF_TICK then the max possible baud rate equals [REF_CLK_FREQ](https://github.com/espressif/esp-idf/blob/master/components/soc/esp32/include/soc/soc.h#L239). I found only one restriction the CONSOLE_UART_BAUDRATE has to be less or equal than the UART clock.

Reply:
@KonstantinKondrashov -

I can tell you with some cursory testing here, if the UART is attached to REF_TICK, a 1MBit baud rate is not reliable on the input side.  (Not surprisingly the output side looks okay in this case.)  If I emit "test" several times into the input of the ESP32 UART at 1MBit while on REF_TICK with the following generator:

```
(for I in $(seq 20);do echo -en test;sleep .0001;done;echo -en "\x03")
```

I get the following from my console on our unit echoed back:

```
testtesttesttesttesttesttesttestY.testesttesttesttesttesttesttest.testtest
```

If the jitter from "sleep" isn't in there, it either completely works or completely fails probably depending on where the asynchronous start bit occurs relative to REF_CLK.  I think there is a limit to the ratio of baud rate to REF_TICK for reliable operation on the input side of the hardware that is not a 1:1 ratio (and which the current software does not capture the limitation of, nor is it in the reference manual).  Try it for yourself.  Of course, baud rates that don't evenly divide into REF_TICK are problematic (such as 230400, 460800, etc) too.

However, that being said, I'll update the patch to use a static assertion.  I'll leave the current check, even though I think it's inadequate for the hardware as it is, until there is a better check.  I'll probably move the check into a macro inside the UART driver header so that it can use the same conditional (and it can be updated later if needed) between the two pieces of code.  Using a macro will support the static assertion.

-- Tim